File: Matlab/src/SPKMeans.m
function [ P, cVs ] = SPKMeans( D, k )
% SPKMeans: Clusters the documents in given document matrix D using the
%   Spherical K-Means Algorithm.
% PARAMETERS:
%   D - the document vector matrix to be clustered.
%   k - number of clusters to partition documents into.
% RETURNS:
%   P - the set of partitions created by the algorithm.
%   cVs - the concept vectors of the returned partition.


    tic;
    
    
    % Set a threshold for change in quality stop condition.
    qThresh = 0.001;
    

    [~, num_docs] = size(D);
    for i = 1:num_docs 
        D(:,i) = D(:,i)/norm(D(:,i));
    end
    
    % Apply TNX or TFN scheme:
    D = txnScheme(D);
    
    
    
    % Create first partitioning (P):
    [wc, dc] = size(D);
    split = floor(dc / k);
    disp(['Split = ' num2str(split)]);

    P = cell(1, k);

    base = 1;
    for i = 1:k % for each partition
        top = base + split - 1;

        if i == k
            top = dc;
        end

        p_size = top - base + 1;
        %P{i} = cell(1, p_size);
        P{i} = zeros(wc, 0);

        %disp(['Base = ' num2str(base)]);
        for j = 1:p_size % for each element in this partition
            v_loc = j + base - 1;
            P{i} = [P{i} D(:, v_loc)];
            %P{i}{j} = D(:, v_loc);
        end

        base = base + split;
        %disp(P{i});
    end
    
    
    
    % Compute concepts for each partition:
    cVs = getConcepts(P, k);
    
    
    % Get initial partition quality:
    q = 0;
    for i = 1:k
        q = q + dot(sum(P{i}, 2), cVs{i});
    end
    dQ = qThresh * 10;
    t = 0;
    % Loop as long as the change in quality is significant:
    while dQ > qThresh && t < 200
        
        % Empty the partitions:
        for i=1:k
            P{i} = zeros(wc, 0);
        end
        
        
        % For each document vector, find closest concept vector of cosine
        %   similarity (smallest dot product):
        for i = 1:dc
            
            closest = 1;
            %cVal = dot(D(:,i), cVs{1}) / (norm(D(:,i)) * norm(cVs{1}));
            cVal = D(:,i)' * cVs{1};
            
            for j=2:k
                % Compute the similarity
                %s = dot(D(:,i), cVs{j}) / (norm(D(:,i)) * norm(cVs{j}));
                s = D(:,i)' * cVs{j};
                
                % If this concept is more similar, choose it instead.
                %   Similarity is closer when s is larger.
                if s > cVal
                    cVal = s;
                    closest = j;
                end
            end
            
            % Assign the vector to the closest partition
            P{closest} = [P{closest} D(:,i)];
        end
        
        
        % Compute the new partition's concepts:
        cVs = getConcepts(P, k);
        
        
        % Compute the quality of the current partitioning:
        lastQ = q;
        q = 0;
        for i = 1:k
            if ~isempty(P{i})
                q = q + dot(sum(P{i}, 2), cVs{i});
            end
        end
        dQ = abs(q - lastQ);
        disp(['q=' num2str(q) ' dQ=' num2str(dQ)]);
        
        
        t = t + 1;
        
    end
    
    
    toc;
    
    
end



function [ concepts ] = getConcepts( partitions, k )
% getConcepts: returns a set of concept vectors, one for each of the
%   given partitions. Concept vectors are indexed in the same order
%   as the partitions given.
% PARAMETERS:
%   partitions - the set of partitions of the document vectors.
%   k - number of partitions.
% RETURNS:
%   concepts - the set of concept vectors for each partition.

    % For all partitions, compute the concept vector of said partition:
    concepts = cell(1, k);
    for i = 1:k
        
        % If partition is empty, make the concept vector a zero vector:
        if isempty(partitions{i})
            [wc, ~] = size(partitions{i});
            concepts{i} = zeros(wc, 1);
            
        else
            % NOTE - mult. by 1/len(P{i}) not in some implementations...
            % Compute mean vector for partition:
            mV_i = (1 / length(partitions{i})) * sum(partitions{i}, 2);

            % Compute concept vector for partition using mean vector:
            concepts{i} = mV_i / norm(mV_i);
        end
        
    end

end



function [ D ] = txnScheme( D )
% txnScheme: Helper function for the SPKMeans algorithm. This function
%   document matrix D and modifies it with the tnx scheme.
% PARAMETERS:
%   D - un-adjusted document matrix D.
% RETURNS:
%   D - adjusted document matrix D.
%
% txn/tfn scheme algorithm (for each document vector x_i):
%      d = number of words in vector
%	f_ji = count of word j in document i
%	t_ji = f_ji
%	 g_j = 1 [txn]
%        = log (n / d_j) [tfn: d_j = # docs containing word j]
%	 s_i = sqrt( sum j=(1, d) { pow((t_ji * g_j), 2) } )
%
%   x_i[j] = t_ji * g_j * s_i


    [~, num_docs] = size(D);
    
    for i = 1:num_docs

        % Calculate s_i for this vector:
        D(:,i) = D(:,i) * sqrt(sum(D(:,i).^2));
        
    % NOTE: for tfn, construct a vector for d_j (# of documents
    %   each word j appears in. Then, multiply:
    % D(:,i) = times(D(:,i), d_vector) * s_i
        
        %vec_i = D(:,i);
        %tempv = vec_i.^2;
        %sum_i = sum(tempv);
        %s_i = sqrt(sum_i);
        %D(:,i) = D(:,i) * s_i
        
        %sum_i = 0;
        %for j = 1:wc
        %    f_ji = D(j, i);
        %    t_ji = f_ji;
        %    g_j = 1;
        %    %sum_i = sum_i + power((t_ji * g_j), 2);
        %    p = (t_ji * g_j);
        %    sum_i = sum_i + (p * p);
        %end
        %s_i = sqrt(sum_i);

        % calculate vector values
        %for j = 1:wc
        %    t_ji = vectors{i}(j); % count of word j in document i
        %    q_j = 1; % constant for txn
        %    vectors{i}(j) = t_ji * q_j * s_i;
        %end

    end
    
end

File: CPP/src/spkmeans.cpp
/* File: spkmeans.cpp
 *
 * Defines the basic SPKMeans functions for the generic SPKMeans class.
 * This class provides a single-threaded implementation of the algorithm,
 * as well as some basic functions useful for the multithreaded versions.
 */

#include "spkmeans.h"

#include "timer.h"
#include "vectors.h"

#include <iostream>
#include <vector>

using namespace std;



// Constructor: initialize variables and document norms.
SPKMeans::SPKMeans(float **doc_matrix_, int k_, int dc_, int wc_)
    : doc_matrix(doc_matrix_), k(k_), dc(dc_), wc(wc_)
{
    // init the doc vector norms
    doc_norms = new float[dc];
    for(int i=0; i<dc; i++)
        doc_norms[i] = vec_norm(doc_matrix[i], wc);

    // default scheme to TXN
    prep_scheme = TXN_SCHEME;

    // initialize optimization to true (optimization will happen)
    optimize = true;
}



// Destructor: clean up document norms array.
SPKMeans::~SPKMeans()
{
    delete[] doc_norms;
}



// Set the current scheme to the given type.
void SPKMeans::setScheme(SPKMeans::Scheme type)
{
    prep_scheme = type;
}



// Disables optimization for testing purposes.
void SPKMeans::disableOptimization()
{
    optimize = false;
}



// Enables optimization if it was disabled.
void SPKMeans::enableOptimization()
{
    optimize = true;
}



// Reports the overall quality and, if optimizing, also displays how many
// clusters have changed.
void SPKMeans::reportQuality(ClusterData *data, float quality, float dQ)
{
    cout << "Quality: " << quality << " (+" << dQ << ")";
    if(optimize) {
        int num_same = 0;
        for (int i=0; i<k; i++)
            if(!(data->changed[i]))
                num_same++;
        cout << " --- " << num_same << " clusters are the same." << endl;
    }
    else
        cout << " --- optimization disabled." << endl;
}



// Reports time data after running the algorithm.
void SPKMeans::reportTime(int iterations, float total_time,
                          float p_time, float c_time, float q_time)
{
    cout << "Done in " << total_time / 1000
         << " seconds after " << iterations << " iterations." << endl;
    float total = p_time + c_time + q_time;
    if(total == 0)
        cout << "No individual time stats available." << endl;
    else {
        cout << "Timers (ms): " << endl
             << "   partitioning [" << p_time << "] ("
                << (p_time/total)*100 << "%)" << endl
             << "   concepts     [" << c_time << "] ("
                << (c_time/total)*100 << "%)" << endl
             << "   quality      [" << q_time << "] ("
                << (q_time/total)*100 << "%)" << endl;
    }
}



// Applies the TXN scheme to each document vector of the given matrix.
// TXN effectively just normalizes each of the document vectors.
// TXN scheme must be set, otherwise this function will do nothing.
void SPKMeans::txnScheme()
{
    if(prep_scheme != SPKMeans::TXN_SCHEME)
        return;

    for(int i=0; i<dc; i++)
        vec_normalize(doc_matrix[i], wc);
}



// Initializes the first partitioning (randomly or otherwise assigned) to
// provide a starting point for the clustering algorithm.
void SPKMeans::initClusters(ClusterData *data)
{
    // choose an initial partitioning
    int split = dc / k;
    cout << "Split = " << split << endl;
    int base = 1;
    for(int i=0; i<k; i++) {
        int top = base + split - 1;
        if(i == k-1)
            top = dc;
        int p_size = top - base + 1;
        for(int j=0; j<p_size; j++)
            data->p_asgns[base-1 + j] = i;
        base = base + split;
    }

    // compute the initial concept vectors
    for(int i=0; i<k; i++)
        data->concepts[i] = computeConcept(data, i);
}



// Returns the total quality of all clusters by summing the qualities of
// each individual cluster. If optimization is enabled, uses cached values
// whenever possible.
float SPKMeans::computeQ(ClusterData *data)
{
    float quality = 0;

    // NOTE - there was a single-cluster computeQ function, re-implement?
    for(int i=0; i<k; i++) {
        if(data->changed[i]) {
            float *sum_p = vec_zeros(wc);
            // add all documents associated with this cluster
            for(int j=0; j<dc; j++) {
                if(data->p_asgns[j] == i) {
                    for(int a=0; a<wc; a++)
                        sum_p[a] += doc_matrix[j][a];
                }
            }
            data->qualities[i] = vec_dot(sum_p, data->concepts[i], wc);
            delete[] sum_p;
        }
        quality += data->qualities[i];
    }

    return quality;
}



// Computes the cosine similarity value of the two given vectors (dv and cv).
float SPKMeans::cosineSimilarity(ClusterData *data, int doc_index, int cIndx)
{
    // TODO - same optimization for concepts? can we cache doc norms better?
    float cnorm = vec_norm(data->concepts[cIndx], wc);
    float dnorm = doc_norms[doc_index];

    // here is where we save time: compute the dot product!
    float dotp = 0;
    for(int i=0; i<(data->docs[doc_index].count); i++) {
        int word = data->docs[doc_index].words[i].index;
        float value = data->docs[doc_index].words[i].value;
        dotp += data->concepts[cIndx][word] * value;
    }
    
    return dotp / (dnorm * cnorm);
}



// Computes the concept vector of the given cluster (by index). The
// cluster documents are accessed using the ClusterData struct, and the
// associated concept vector will be allocated and populated.
float* SPKMeans::computeConcept(ClusterData *data, int cIndx)
{
    // create the concept vector and initialize it to 0
    float *concept = new float[wc];
    for(int i=0; i<wc; i++)
        concept[i] = 0;

    // find documents associated w/ this cluster, and sum them
    for(int i=0; i<dc; i++) { // for each document
        if(data->p_asgns[i] == cIndx) { // if doc in cluster
            for(int j=0; j<wc; j++) // add words to concept
                concept[j] += doc_matrix[i][j];
        }
    }

    // normalize the concept vector and return it
    //vec_divide(concept, wc, wc);
    vec_normalize(concept, wc);
    return concept;
}



// Runs the spherical k-means algorithm on the given sparse matrix D and
// clusters the data into k clusters. Non-parallel (standard) version.
ClusterData* SPKMeans::runSPKMeans()
{
    // keep track of the run time for this algorithm
    Timer timer;
    timer.start();

    // keep track of all individual component times for analysis
    Timer ptimer;
    Timer ctimer;
    Timer qtimer;

    // apply the TXN scheme on the document vectors (normalize them)
    txnScheme();

    // initialize the data arrays; keep track of the arrays locally
    ClusterData *data = new ClusterData(k, dc, wc, doc_matrix);
    float **concepts = data->concepts;
    bool *changed = data->changed;
    float *cosines = data->cosine_similarities;

    // compute initial partitioning, concepts, and quality
    initClusters(data);
    float quality = computeQ(data);
    cout << "Initial quality: " << quality << endl;
    

    // do spherical k-means loop
    float dQ = Q_THRESHOLD * 10;
    int iterations = 0;
    while(dQ > Q_THRESHOLD) {
        iterations++;

        // TODO - temporary (testing) -----------------------------------------
        /*if(iterations > 1) {
            cout << "New version." << endl;
            ptimer.start();
            float averagePriority = data->getAverageMovedPriority();
            for(int i=0; i<dc; i++) {
                if(data->doc_priorities[i] > averagePriority) {
                    int cIndx = 0;
                    if(changed[0])
                        cosines[i*k] = cosineSimilarity(data, i, 0);
                    for(int j=1; j<k; j++) {
                        if(changed[j])
                            cosines[i*k + j] = cosineSimilarity(data, i, j);
                        if(cosines[i*k + j] > cosines[i*k + cIndx])
                            cIndx = j;
                    }
                    // compute the priority heuristic and assign the document
                    float priority = 1 - cosines[i*k + data->p_asgns[i]];
                    data->assignCluster(i, cIndx, priority);
                }
            }
        } else {*/
        // --------------------------------------------------------------------

        // compute new clusters based on old concept vectors
        ptimer.start();

        // TODO - temporary testing for empty clusters
        bool has_docs[k];
        for(int i=0; i<k; i++)
            has_docs[i] = false;

        for(int i=0; i<dc; i++) {
            // only update cosine similarities if cluster has changed
            int cIndx = 0;
            if(changed[0])
                cosines[i*k] = cosineSimilarity(data, i, 0);
            for(int j=1; j<k; j++) {
                if(changed[j])
                    cosines[i*k + j] = cosineSimilarity(data, i, j);
                if(cosines[i*k + j] > cosines[i*k + cIndx])
                    cIndx = j;
            }
            // compute the priority heuristic and assign the document
            //float priority = 1 - cosines[i*k + data->p_asgns[i]];
            data->assignCluster(i, cIndx);//, priority);
            has_docs[cIndx] = true;
        }//}

        // TODO - temporary testing for empty clusters:
        for(int i=0; i<k; i++) {
            if(!has_docs[i])
                cout << "Cluster " << i << " is empty!" << endl;
        }

        ptimer.stop();

        // TODO - temporary (testing) -----------------------------------------
        // report priorities and number of documents that moved
        /*cout << "Number of documents moved: " << data->num_moved << endl;
        cout << "Average document priority: "
             << data->getAveragePriority() << endl;
        cout << "   Average moved priority: "
             << data->getAverageMovedPriority() << endl;
        cout << "  Average stayed priority: "
             << data->getAverageStayPriority() << endl;*/
        // --------------------------------------------------------------------

        // update which clusters changed since last time, then swap pointers
        if(optimize)
            data->findChangedClusters();
        data->applyAssignments();

        // compute new concept vectors
        ctimer.start();
        for(int i=0; i<k; i++) {
            if(changed[i]) {
                delete[] concepts[i];
                concepts[i] = computeConcept(data, i);
            }
        }
        ctimer.stop();

        // compute quality of new partitioning
        qtimer.start();
        float n_quality = computeQ(data);
        dQ = n_quality - quality;
        quality = n_quality;
        qtimer.stop();

        // report the quality of the current partitioning
        reportQuality(data, quality, dQ);
    }


    // report runtime statistics
    timer.stop();
    reportTime(iterations, timer.get(), ptimer.get(), ctimer.get(),
               qtimer.get());

    // return the resulting clusters and concepts in the ClusterData struct
    return data;
}
